File: misc/PCF8574/bcm2835/pcf8574.c
#include <bcm2835.h>  
#include <stdio.h>

#define beep_on  i2c_writeByte(0x7F & i2c_readByte())
#define beep_off i2c_writeByte(0x80 | i2c_readByte())
#define led_off   i2c_writeByte(0x10 | i2c_readByte())
#define led_on  i2c_writeByte(0xEF & i2c_readByte())
void i2c_writeByte(char byte)
{
	char buf[] = {byte};
	bcm2835_i2c_write(buf,1);
}
char i2c_readByte()
{
	char buf[1];
	bcm2835_i2c_read(buf,1);
	return buf[0];
}

int main(int argc, char **argv)  
{  
	char value;

	if (!bcm2835_init())return 1;  
	bcm2835_i2c_begin(); 
	bcm2835_i2c_setSlaveAddress(0x20);  
    bcm2835_i2c_set_baudrate(10000);  
    printf("PCF8574 Test Program !!!\n"); 

    while(1)  
    {  
		i2c_writeByte(0x0F | i2c_readByte());
		value = i2c_readByte() | 0xF0;
		if(value != 0xFF)
		{
			beep_on;
			led_on;
			switch(value)
			{	
				case 0xFE:
					printf("left\n");break;	
				case 0xFD:
					printf("up\n");	break;
				case 0xFB:
					printf("dowm\n");break;	
				case 0xF7:
					printf("right\n");break;
				default :
					printf("unknow\n");
			}
			while(value != 0xFF)
			{
				i2c_writeByte(0x0F | i2c_readByte());
				value = i2c_readByte() | 0xF0;
				bcm2835_delay(10);
			}
			beep_off;
			led_off;
		}
		bcm2835_delay(10);	
	}           
    bcm2835_i2c_end();  
    bcm2835_close();  
    return 0;  
}  

File: misc2/PCF8574/python/pcf8574.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
import RPi.GPIO as GPIO
import smbus
import time

address = 0x20

def beep_on():
	bus.write_byte(address,0x7F&bus.read_byte(address))
def beep_off():
	bus.write_byte(address,0x80|bus.read_byte(address))
def led_off():
	bus.write_byte(address,0x10|bus.read_byte(address))
def led_on():
	bus.write_byte(address,0xEF&bus.read_byte(address))

bus = smbus.SMBus(1)
print("PCF8574 Test Program !!!")
while True:
	bus.write_byte(address,0x0F|bus.read_byte(address))
	value = bus.read_byte(address) | 0xF0
	if value != 0xFF:
		led_on()
		beep_on()
		if (value | 0xFE) != 0xFF:
			print("left")
		elif (value | 0xFD) != 0xFF:
			print("up")
		elif (value | 0xFB) != 0xFF:
			print("dowm")
		else :
			print("right")
		while value != 0xFF:
			bus.write_byte(address,0x0F|bus.read_byte(address))
			value = bus.read_byte(address) | 0xF0
			time.sleep(0.01)
		beep_off()
		led_off()
	time.sleep(0.1)


